File: pipeline/highlighter_pipeline/supabase_client.py
import json
import mimetypes
from datetime import datetime, timezone
from pathlib import Path
from typing import Any
from urllib.error import HTTPError, URLError
from urllib.parse import quote, urlencode
from urllib.request import Request, urlopen

from .config import required_any_env, required_env


class SupabaseClient:
    def __init__(self) -> None:
        self.base_url = required_env("SUPABASE_URL").rstrip("/")
        self.key = required_any_env(["SUPABASE_SERVICE_ROLE_KEY", "SUPABASE_SECRET_KEY"])

    def create_project(
        self,
        *,
        name: str,
        source_type: str,
        source_url: str,
        status: str = "ingesting",
        metadata: dict[str, Any] | None = None,
    ) -> str:
        rows = self._request(
            "POST",
            "projects",
            query={"select": "id"},
            body={
                "name": name,
                "source_type": source_type,
                "source_url": source_url,
                "status": status,
                "metadata": metadata or {},
            },
            prefer="return=representation",
        )
        return rows[0]["id"]

    def set_project_status(
        self,
        project_id: str,
        *,
        status: str,
        metadata: dict[str, Any] | None = None,
        error: str | None = None,
    ) -> None:
        body: dict[str, Any] = {"status": status}
        if metadata is not None:
            body["metadata"] = metadata
        if error is not None:
            body["error"] = error

        self._request(
            "PATCH",
            "projects",
            query={"id": f"eq.{project_id}"},
            body=body,
            prefer="return=minimal",
        )

    def update_project_status_guarded(
        self,
        project_id: str,
        *,
        status: str,
        when_status_in: list[str],
        metadata: dict[str, Any] | None = None,
        error: str | None = None,
    ) -> dict[str, Any] | None:
        """PATCH the project only while its status is one of when_status_in, so the
        worker never overwrites a terminal/cancellation status the router wrote.
        Returns the updated row, or None when the guard matched nothing."""
        body: dict[str, Any] = {"status": status}
        if metadata is not None:
            body["metadata"] = metadata
        if error is not None:
            body["error"] = error

        rows = self._request(
            "PATCH",
            "projects",
            query={
                "id": f"eq.{project_id}",
                "status": f"in.({','.join(when_status_in)})",
            },
            body=body,
            prefer="return=representation",
        )
        return rows[0] if rows else None

    def get_project_status(self, project_id: str) -> str | None:
        rows = self._request(
            "GET",
            "projects",
            query={"id": f"eq.{project_id}", "select": "status"},
        )
        return rows[0]["status"] if rows else None

    def insert_transcript_chunk(
        self,
        *,
        project_id: str,
        chunk_index: int,
        start_seconds: float,
        end_seconds: float,
        transcript: str,
        words: list[dict[str, Any]],
        metadata: dict[str, Any] | None = None,
    ) -> None:
        self._request(
            "POST",
            "transcript_chunks",
            body={
                "project_id": project_id,
                "chunk_index": chunk_index,
                "start_seconds": start_seconds,
                "end_seconds": end_seconds,
                "transcript": transcript,
                "words": words,
                "metadata": metadata or {},
            },
            prefer="return=minimal",
        )

    def insert_clip(
        self,
        *,
        project_id: str,
        title: str,
        start_seconds: float,
        end_seconds: float,
        description: str | None = None,
        score: float | None = None,
        video_url: str | None = None,
        status: str | None = None,
        metadata: dict[str, Any] | None = None,
    ) -> None:
        body: dict[str, Any] = {
            "project_id": project_id,
            "title": title,
            "start_seconds": start_seconds,
            "end_seconds": end_seconds,
            "description": description,
            "score": score,
            "video_url": video_url,
            "metadata": metadata or {},
        }
        if status is not None:
            body["status"] = status

        self._request(
            "POST",
            "clips",
            body=body,
            prefer="return=minimal",
        )

    def get_project(self, project_id: str) -> dict[str, Any]:
        rows = self._request(
            "GET",
            "projects",
            query={"id": f"eq.{project_id}", "select": "*"},
        )
        if not rows:
            raise RuntimeError(f"Project not found: {project_id}")
        return rows[0]

    def pending_cleanup_jobs(self, *, limit: int = 100) -> list[dict[str, Any]]:
        return (
            self._request(
                "GET",
                "media_cleanup_jobs",
                query={"status": "eq.pending", "order": "created_at", "limit": str(limit)},
            )
            or []
        )

    def finish_cleanup_job(self, job_id: str, *, status: str, error: str | None = None) -> None:
        body: dict[str, Any] = {
            "status": status,
            "completed_at": datetime.now(timezone.utc).isoformat(),
        }
        if error is not None:
            body["last_error"] = error

        self._request(
            "PATCH",
            "media_cleanup_jobs",
            query={"id": f"eq.{job_id}"},
            body=body,
            prefer="return=minimal",
        )

    def bump_cleanup_attempts(self, job_id: str, attempts: int) -> None:
        self._request(
            "PATCH",
            "media_cleanup_jobs",
            query={"id": f"eq.{job_id}"},
            body={"attempts": attempts},
            prefer="return=minimal",
        )

    def delete_storage_object(self, *, bucket: str, key: str) -> None:
        storage_path = f"{quote(bucket, safe='')}/{quote(key, safe='/')}"
        request = Request(
            f"{self.base_url}/storage/v1/object/{storage_path}",
            headers={
                "apikey": self.key,
                "Authorization": f"Bearer {self.key}",
            },
            method="DELETE",
        )

        try:
            with urlopen(request, timeout=60) as response:
                response.read()
        except HTTPError as exc:
            if exc.code == 404:
                return  # already gone — deletion is idempotent
            message = exc.read().decode("utf-8", errors="replace")
            raise RuntimeError(f"Supabase storage delete failed: {message}") from exc
        except URLError as exc:
            raise RuntimeError(f"Supabase storage delete failed: {exc.reason}") from exc

    def upload_storage_object(self, *, bucket: str, key: str, path: Path) -> str:
        content_type = mimetypes.guess_type(path.name)[0] or "application/octet-stream"
        storage_path = f"{quote(bucket, safe='')}/{quote(key, safe='/')}"
        request = Request(
            f"{self.base_url}/storage/v1/object/{storage_path}",
            data=path.read_bytes(),
            headers={
                "apikey": self.key,
                "Authorization": f"Bearer {self.key}",
                "Content-Type": content_type,
                "x-upsert": "true",
            },
            method="POST",
        )

        try:
            with urlopen(request, timeout=120) as response:
                response.read()
        except HTTPError as exc:
            message = exc.read().decode("utf-8", errors="replace")
            raise RuntimeError(f"Supabase storage upload failed: {message}") from exc
        except URLError as exc:
            raise RuntimeError(f"Supabase storage upload failed: {exc.reason}") from exc

        return self.public_storage_url(bucket=bucket, key=key)

    def public_storage_url(self, *, bucket: str, key: str) -> str:
        storage_path = f"{quote(bucket, safe='')}/{quote(key, safe='/')}"
        return f"{self.base_url}/storage/v1/object/public/{storage_path}"

    def _request(
        self,
        method: str,
        table: str,
        *,
        query: dict[str, str] | None = None,
        body: dict[str, Any] | None = None,
        prefer: str | None = None,
    ) -> Any:
        query_string = f"?{urlencode(query)}" if query else ""
        data = json.dumps(body).encode("utf-8") if body is not None else None
        headers = {
            "apikey": self.key,
            "Authorization": f"Bearer {self.key}",
            "Content-Type": "application/json",
        }
        if prefer:
            headers["Prefer"] = prefer

        request = Request(
            f"{self.base_url}/rest/v1/{table}{query_string}",
            data=data,
            headers=headers,
            method=method,
        )

        try:
            with urlopen(request, timeout=30) as response:
                response_body = response.read().decode("utf-8")
        except HTTPError as exc:
            message = exc.read().decode("utf-8", errors="replace")
            raise RuntimeError(f"Supabase {method} {table} failed: {message}") from exc
        except URLError as exc:
            raise RuntimeError(f"Supabase {method} {table} failed: {exc.reason}") from exc

        if not response_body:
            return None
        return json.loads(response_body)

File: pipeline-dotnet/src/Highlighter.Pipeline/SupabaseClient.cs
using System.Globalization;
using System.Text;
using System.Text.Json.Nodes;

namespace Highlighter.Pipeline;

/// <summary>Port of highlighter_pipeline/supabase_client.py: PostgREST + Storage
/// over raw HTTP.</summary>
public class SupabaseClient
{
    private static readonly HttpClient Http = new()
    {
        Timeout = System.Threading.Timeout.InfiniteTimeSpan,
    };

    public string BaseUrl { get; }
    public string Key { get; }

    public SupabaseClient()
    {
        BaseUrl = Config.RequiredEnv("SUPABASE_URL").TrimEnd('/');
        Key = Config.RequiredAnyEnv(new[] { "SUPABASE_SERVICE_ROLE_KEY", "SUPABASE_SECRET_KEY" });
    }

    public string CreateProject(
        string name,
        string sourceType,
        string sourceUrl,
        string status = "ingesting",
        JsonObject? metadata = null)
    {
        var rows = Request(
            "POST",
            "projects",
            query: new() { ["select"] = "id" },
            body: new JsonObject
            {
                ["name"] = name,
                ["source_type"] = sourceType,
                ["source_url"] = sourceUrl,
                ["status"] = status,
                ["metadata"] = metadata is null ? new JsonObject() : JsonUtil.CloneObj(metadata),
            },
            prefer: "return=representation");
        return JsonUtil.Str((rows as JsonArray)?[0]?["id"]);
    }

    public void SetProjectStatus(
        string projectId, string status, JsonObject? metadata = null, string? error = null)
    {
        var body = new JsonObject { ["status"] = status };
        if (metadata is not null) body["metadata"] = JsonUtil.CloneObj(metadata);
        if (error is not null) body["error"] = error;

        Request(
            "PATCH",
            "projects",
            query: new() { ["id"] = $"eq.{projectId}" },
            body: body,
            prefer: "return=minimal");
    }

    /// <summary>PATCH the project only while its status is one of whenStatusIn, so the
    /// worker never overwrites a terminal/cancellation status the router wrote.
    /// Returns the updated row, or null when the guard matched nothing.</summary>
    public JsonObject? UpdateProjectStatusGuarded(
        string projectId,
        string status,
        IReadOnlyList<string> whenStatusIn,
        JsonObject? metadata = null,
        string? error = null)
    {
        var body = new JsonObject { ["status"] = status };
        if (metadata is not null) body["metadata"] = JsonUtil.CloneObj(metadata);
        if (error is not null) body["error"] = error;

        var rows = Request(
            "PATCH",
            "projects",
            query: new()
            {
                ["id"] = $"eq.{projectId}",
                ["status"] = $"in.({string.Join(",", whenStatusIn)})",
            },
            body: body,
            prefer: "return=representation");
        return rows is JsonArray { Count: > 0 } array ? array[0] as JsonObject : null;
    }

    /// <summary>PATCH only the metadata jsonb, leaving status untouched.</summary>
    public void UpdateProjectMetadata(string projectId, JsonObject metadata)
    {
        Request(
            "PATCH",
            "projects",
            query: new() { ["id"] = $"eq.{projectId}" },
            body: new JsonObject { ["metadata"] = JsonUtil.CloneObj(metadata) },
            prefer: "return=minimal");
    }

    public string? GetProjectStatus(string projectId)
    {
        var rows = Request(
            "GET",
            "projects",
            query: new() { ["id"] = $"eq.{projectId}", ["select"] = "status" });
        return rows is JsonArray { Count: > 0 } array
            ? JsonUtil.StrOrNull(array[0]?["status"])
            : null;
    }

    public void InsertTranscriptChunk(
        string projectId,
        int chunkIndex,
        double startSeconds,
        double endSeconds,
        string transcript,
        JsonArray words,
        JsonObject? metadata = null)
    {
        Request(
            "POST",
            "transcript_chunks",
            body: new JsonObject
            {
                ["project_id"] = projectId,
                ["chunk_index"] = chunkIndex,
                ["start_seconds"] = startSeconds,
                ["end_seconds"] = endSeconds,
                ["transcript"] = transcript,
                ["words"] = (JsonArray)words.DeepClone(),
                ["metadata"] = metadata is null ? new JsonObject() : JsonUtil.CloneObj(metadata),
            },
            prefer: "return=minimal");
    }

    public void InsertClip(
        string projectId,
        string title,
        double startSeconds,
        double endSeconds,
        string? description = null,
        JsonNode? score = null,
        string? videoUrl = null,
        string? verticalUrl = null,
        string? status = null,
        JsonObject? metadata = null)
    {
        var body = new JsonObject
        {
            ["project_id"] = projectId,
            ["title"] = title,
            ["start_seconds"] = startSeconds,
            ["end_seconds"] = endSeconds,
            ["description"] = description,
            ["score"] = JsonUtil.C(score),
            ["video_url"] = videoUrl,
            ["vertical_url"] = verticalUrl,
            ["metadata"] = metadata is null ? new JsonObject() : JsonUtil.CloneObj(metadata),
        };
        if (status is not null) body["status"] = status;

        Request("POST", "clips", body: body, prefer: "return=minimal");
    }

    public void InsertLongformEdit(
        string projectId,
        int version,
        string status = "rendered",
        string? videoUrl = null,
        string? thumbnailUrl = null,
        double? durationSeconds = null,
        JsonArray? segments = null,
        JsonObject? editor = null,
        JsonObject? revision = null,
        JsonObject? metadata = null)
    {
        Request(
            "POST",
            "longform_edits",
            body: new JsonObject
            {
                ["project_id"] = projectId,
                ["version"] = version,
                ["status"] = status,
                ["video_url"] = videoUrl,
                ["thumbnail_url"] = thumbnailUrl,
                ["duration_seconds"] = durationSeconds,
                ["segments"] = segments is null ? new JsonArray() : (JsonArray)segments.DeepClone(),
                ["editor"] = editor is null ? null : JsonUtil.CloneObj(editor),
                ["revision"] = revision is null ? null : JsonUtil.CloneObj(revision),
                ["metadata"] = metadata is null ? new JsonObject() : JsonUtil.CloneObj(metadata),
            },
            prefer: "return=minimal");
    }

    public JsonObject GetProject(string projectId)
    {
        var rows = Request(
            "GET",
            "projects",
            query: new() { ["id"] = $"eq.{projectId}", ["select"] = "*" });
        if (rows is not JsonArray { Count: > 0 } array || array[0] is not JsonObject row)
            throw new PipelineError($"Project not found: {projectId}");
        return row;
    }

    public List<JsonObject> PendingCleanupJobs(int limit = 100)
    {
        var rows = Request(
            "GET",
            "media_cleanup_jobs",
            query: new()
            {
                ["status"] = "eq.pending",
                ["order"] = "created_at",
                ["limit"] = limit.ToString(CultureInfo.InvariantCulture),
            });
        return JsonUtil.Objects(rows);
    }

    public void FinishCleanupJob(string jobId, string status, string? error = null)
    {
        var body = new JsonObject
        {
            ["status"] = status,
            ["completed_at"] = DateTime.UtcNow.ToString(
                "yyyy-MM-dd'T'HH:mm:ss.ffffff+00:00", CultureInfo.InvariantCulture),
        };
        if (error is not null) body["last_error"] = error;

        Request(
            "PATCH",
            "media_cleanup_jobs",
            query: new() { ["id"] = $"eq.{jobId}" },
            body: body,
            prefer: "return=minimal");
    }

    public void BumpCleanupAttempts(string jobId, int attempts)
    {
        Request(
            "PATCH",
            "media_cleanup_jobs",
            query: new() { ["id"] = $"eq.{jobId}" },
            body: new JsonObject { ["attempts"] = attempts },
            prefer: "return=minimal");
    }

    public void DeleteStorageObject(string bucket, string key)
    {
        var storagePath = $"{Uri.EscapeDataString(bucket)}/{QuoteKey(key)}";
        using var request = new HttpRequestMessage(
            HttpMethod.Delete, $"{BaseUrl}/storage/v1/object/{storagePath}");
        request.Headers.TryAddWithoutValidation("apikey", Key);
        request.Headers.TryAddWithoutValidation("Authorization", $"Bearer {Key}");

        try
        {
            using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(60));
            using var response = Http.Send(request, HttpCompletionOption.ResponseContentRead, cts.Token);
            if (response.IsSuccessStatusCode) return;
            if ((int)response.StatusCode == 404) return; // already gone — deletion is idempotent
            var message = response.Content.ReadAsStringAsync().GetAwaiter().GetResult();
            throw new PipelineError($"Supabase storage delete failed: {message}");
        }
        catch (HttpRequestException exc)
        {
            throw new PipelineError($"Supabase storage delete failed: {exc.Message}", exc);
        }
        catch (TaskCanceledException exc)
        {
            throw new PipelineError("Supabase storage delete failed: timed out", exc);
        }
    }

    public string UploadStorageObject(string bucket, string key, string path)
    {
        var contentType = MimeTypes.Guess(path) ?? "application/octet-stream";
        var storagePath = $"{Uri.EscapeDataString(bucket)}/{QuoteKey(key)}";
        using var request = new HttpRequestMessage(
            HttpMethod.Post, $"{BaseUrl}/storage/v1/object/{storagePath}")
        {
            Content = new ByteArrayContent(File.ReadAllBytes(path)),
        };
        request.Headers.TryAddWithoutValidation("apikey", Key);
        request.Headers.TryAddWithoutValidation("Authorization", $"Bearer {Key}");
        request.Content.Headers.TryAddWithoutValidation("Content-Type", contentType);
        request.Headers.TryAddWithoutValidation("x-upsert", "true");

        try
        {
            using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(120));
            using var response = Http.Send(request, HttpCompletionOption.ResponseContentRead, cts.Token);
            if (!response.IsSuccessStatusCode)
            {
                var message = response.Content.ReadAsStringAsync().GetAwaiter().GetResult();
                throw new PipelineError($"Supabase storage upload failed: {message}");
            }
        }
        catch (HttpRequestException exc)
        {
            throw new PipelineError($"Supabase storage upload failed: {exc.Message}", exc);
        }
        catch (TaskCanceledException exc)
        {
            throw new PipelineError("Supabase storage upload failed: timed out", exc);
        }

        return PublicStorageUrl(bucket: bucket, key: key);
    }

    public string PublicStorageUrl(string bucket, string key)
    {
        var storagePath = $"{Uri.EscapeDataString(bucket)}/{QuoteKey(key)}";
        return $"{BaseUrl}/storage/v1/object/public/{storagePath}";
    }

    /// <summary>quote(key, safe='/')</summary>
    private static string QuoteKey(string key) =>
        string.Join("/", key.Split('/').Select(Uri.EscapeDataString));

    private JsonNode? Request(
        string method,
        string table,
        Dictionary<string, string>? query = null,
        JsonObject? body = null,
        string? prefer = null)
    {
        var queryString = query is { Count: > 0 }
            ? "?" + string.Join("&", query.Select(pair =>
                $"{Uri.EscapeDataString(pair.Key)}={Uri.EscapeDataString(pair.Value)}"))
            : "";

        using var request = new HttpRequestMessage(
            new HttpMethod(method), $"{BaseUrl}/rest/v1/{table}{queryString}");
        if (body is not null)
            request.Content = new StringContent(JsonUtil.Dumps(body), Encoding.UTF8, "application/json");
        request.Headers.TryAddWithoutValidation("apikey", Key);
        request.Headers.TryAddWithoutValidation("Authorization", $"Bearer {Key}");
        if (prefer is not null) request.Headers.TryAddWithoutValidation("Prefer", prefer);

        string responseBody;
        try
        {
            using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(30));
            using var response = Http.Send(request, HttpCompletionOption.ResponseContentRead, cts.Token);
            responseBody = response.Content.ReadAsStringAsync().GetAwaiter().GetResult();
            if (!response.IsSuccessStatusCode)
                throw new PipelineError($"Supabase {method} {table} failed: {responseBody}");
        }
        catch (HttpRequestException exc)
        {
            throw new PipelineError($"Supabase {method} {table} failed: {exc.Message}", exc);
        }
        catch (TaskCanceledException exc)
        {
            throw new PipelineError($"Supabase {method} {table} failed: timed out", exc);
        }

        if (string.IsNullOrEmpty(responseBody)) return null;
        return JsonUtil.Parse(responseBody);
    }
}

/// <summary>mimetypes.guess_type, reduced to the types this pipeline uploads.</summary>
public static class MimeTypes
{
    private static readonly Dictionary<string, string> Map = new(StringComparer.OrdinalIgnoreCase)
    {
        [".mp4"] = "video/mp4",
        [".jpg"] = "image/jpeg",
        [".jpeg"] = "image/jpeg",
        [".png"] = "image/png",
        [".gif"] = "image/gif",
        [".wav"] = "audio/x-wav",
        [".mp3"] = "audio/mpeg",
        [".ts"] = "video/mp2t",
        [".json"] = "application/json",
        [".txt"] = "text/plain",
        [".html"] = "text/html",
    };

    public static string? Guess(string path) =>
        Map.TryGetValue(Path.GetExtension(path), out var contentType) ? contentType : null;
}
